src/fn/edit-contacts.js

```diff
@@ -34,16 +52,29 @@ const execute = () => {
         const nameParts = fs.path.basename(csv).split('.');
         const prefix = nameParts[0];
         switch(prefix) {
-          case 'contact':  return processDocuments(csv, getIDs(csv, prefix), db, args);
-          case 'users': return processUsers(csv, getIDs(csv, prefix), db, args);
+          case 'contact':  return processDocuments('contact', csv, getIDs(csv, prefix), db, args);
+          case 'users': return processDocuments('user', csv, getIDs(csv, prefix), db, args);
           default: throw new Error(`Unrecognised CSV type ${prefix} for file ${csv}`);
         }
       })
       .then(docs => addToModel(csv, docs)),
       Promise.resolve())
 
     .then(() => model.exclusions.forEach(toDocs.removeExcludedField))
-    .then(() => Promise.all(Object.values(model.docs).map(saveJsonDoc)));
+    .then(() => Promise.all(Object.values(model.docs).map(doc => {
+      const saveDoc = saveJsonDoc(doc, args, overwriteAllFiles);
+
+      if(saveDoc === DOC_EXISTS) {
+        const userSelection = userPrompt.keyInSelect();
+
+        if (userSelection === undefined || userSelection === 2) {
+          throw new Error('User canceled the action.');
+        }
+        overwriteAllFiles = (userSelection === 1);
+
+        saveJsonDoc(doc, args, overwriteAllFiles);
+      }
+    })));
```

Comment:
The `Promise.all` here in the original code received promises for each `saveJsonDoc` call. In the change here though, the `return` is absent, so the `map` function is just going to return an array of `undefined` and we'll have an async bug. Adding a `return` on like 75 will cover some cases, but since that's in a branch based on 67, the "else" branch needs a return as well.

I propose a change that will cover that async behavior and also make this a bit more readable. The `if(saveDoc === DOC_EXISTS)` check can be moved to a `filter` as part of this chain. That will help to make this new function strictly about what it does do, rather than having to include a filter for what it doesn't do. Once that's been done, the function can more easily be pulled out into a named function (similar to how `saveJsonDoc` is) and then called in the chain here.

The big benefit with that is a reader can just scan over this area of the code and see the chain and nice names to get the gist of what's going on, without having to dive into all the details to reason about it.

Reply:
Very good catch! I have factored out overwrite file check from the `saveJsonDoc` function. It works better than having a filter because we might need to prompt user input on each doc.